mono/mini/mini-llvm.c

```diff
@@ -3946,70 +3937,90 @@ mono_protected_region_insert (EmitContext *ctx, MonoProtectedRegion **out, MonoE
 	// We have to statically allocate the switches that we do when doing
 	// exception handling. Each function gets visited once. This means that we
 	// need what is essentially an interval tree. When looking at a node we need to know where
-	// the parent is and which clauses are handled by that interval and not any children.
-	// Each clause group needs a parent and it needs a set of children
+	// the parent is and which clauses are handled by that interval and not any catch_clauses.
+	// Each clause group needs a parent and it needs a set of catch_clauses
 
 	MonoExceptionClause *ex = mono_protected_region_elem (*out);
 	MonoProtectedRegion *val = *out;
 
 	if (CLAUSE_START (input) < CLAUSE_START (ex) || CLAUSE_END(ex) > CLAUSE_END (input)) {
+		g_assert_not_reached ();
 		// Input contains current tree as sub-tree
 		MonoProtectedRegion *root = mono_mempool_alloc0 (ctx->cfg->mempool, sizeof (MonoProtectedRegion));
 		mono_protected_region_add (ctx, root, input);
-		root->subintervals = g_slist_prepend_mempool (ctx->mempool, root->subintervals, val);
+		root->nested = g_slist_prepend_mempool (ctx->mempool, root->nested, val);
 		*out = root;
 
 		return TRUE;
-	} else if (CLAUSE_START (input) > CLAUSE_START (ex) || CLAUSE_END(ex) < CLAUSE_END (input)) {
-		// If the element is a child, and at this level
+	} else if (CLAUSE_START (input) == CLAUSE_START (ex) || CLAUSE_END(ex) == CLAUSE_END (input)) {
+		// If the element is mutually protecting, and at this level
 		mono_protected_region_add (ctx, val, input);
 
 		return TRUE;
-	} else if (CLAUSE_START (input) > CLAUSE_START (ex) || CLAUSE_END(ex) != CLAUSE_END (input)) {
-		// If the element is a child, and not at this level
-		for (GSList *curr = val->subintervals; curr; curr = curr->next) {
-			MonoExceptionClause *cast = mono_protected_region_elem (curr->data);
-			if (CLAUSE_START (input) <= CLAUSE_START (cast) || CLAUSE_END(cast) >= CLAUSE_END (input))
-				return mono_protected_region_insert (ctx, (MonoProtectedRegion **)&curr->data, input);
+	} else if (CLAUSE_START (input) >= CLAUSE_START (ex) || CLAUSE_END(ex) != CLAUSE_END (input)) {
```

Comment:
Since you always return, none of these needs to be an `else`, saving some nesting; ditto the returns and break in this branch.